### PR TITLE
Added 3 passes for the ornithopter airstrike

### PR DESCRIPTION
**Pull request by Mailaender**
This might need some balancing adjustments as well.

### CODE REVIEW

**Comment by sorcerer86pt** [comment]
Okay, checked it and it has 3 passes

[ornithopter passes](https://drive.google.com/file/d/1oQY70KfqXfYhKviQn-sGAQj6uhhyODHl/view?usp=sharing)

**Review by sorcerer86pt — DISMISSED** [review]

**Comment by pchote** [comment]
@sorcerer86pt's video shows that this appears to be missing a key detail from the ornithopter strike: they should circle by ~270 degrees, such that the second strike is at right angles to the first, and the third at ~180 degrees, so they fly off the map in roughly the same direction that they flew in.

**Comment by sorcerer86pt** [comment]
Didn't know that aspect ( didn't play dune 2000). I just counted the strike
phases.


Em 21/03/2020 à(s) 8:05 da tarde, Paul Chote <notifications@github.com>
escreveu:

@sorcerer86pt <https://github.com/sorcerer86pt>'s video shows that this
appears to be missing a key detail from the ornithopter strike: they should
circle by ~270 degrees, such that the second strike is at right angles to
the third, and they fly off the map in roughly the same direction that they
flew in.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/OpenRA/OpenRA/pull/17827#issuecomment-602096661>, or
unsubscribe

**Comment by penev92** [comment]
In the video that I linked to in the original issue they seem to exit the map at about 60° to their entry trajectory - https://youtu.be/dmflM70vtZI?t=36 (timestamped).
Watching the video @sorcerer86pt provided, they seem a bit too eager to turn, barely managing to turn towards their target because they start turning too early, but one could still take this as an improvement over the original where they would take their time and circle half the map, giving the opponent all the time to repair affected structures and shoot them down.

Here is another timestamped video - https://youtu.be/_t3Tuuoy_P4?list=PLw2L04beHwWBbjNh8AnY2OTbHKb5XeDlO&t=577
The angles seem to match the first one, but the formation is broken. Also it seems they don't bomb in a straight line here as well, much like the video from @sorcerer86pt (and unlike my first video).
And another one: https://youtu.be/_t3Tuuoy_P4?list=PLw2L04beHwWBbjNh8AnY2OTbHKb5XeDlO&t=3335

**Comment by penev92** [comment]
Considering how poorly Ornithopters seem to fare in the original, we need to decide how close to that we want to get with this.

**Comment by penev92** [comment]
I think it's a pretty valid point to say that
 - The formation we use (like for Badgers in RA and A10s in TD) is nicer than the original line formation.
 - The decreased turning arcs do make this support power a bit more potent, by both decreasing the total time over which the damage is dealt and preventing the Ornithopters from straying too far into enemy AA fire.
 - The angles at which the ornithopters hit are hardly crucial.
 - The triangle formation plus the turning make for some weird formation reordering, but having them keep formation while turning would require each one to individually adjust its movement speed, which is unrealistic at this point due to engine limitations.

Here is another video (in case the previous ones weren't enough), this time made by me: https://gfycat.com/firmbetterkite
Also the Ornithopters seem to come into the map from where the High Tech Factory is.

We do need to settle on a specification for how we want all of this to work.

**Comment by pchote** [comment]
> which is unrealistic at this point due to engine limitations.

No it isn't. It would be straightforward to create a new fly-type activity that allows the aircraft to follow a path with a constant offset to keep formation. Someone just needs to write the code.

**Comment by sorcerer86pt** [comment]
we could define a steering behaviour of leader- follower in formation ,
where 1 is the leader that you define the path, and the other 2 , the
followers, follow the leader path with a given offset. This also allows
formations for units.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs [hunk 1]
```diff
@@ -48,6 +48,12 @@ public class AirstrikePowerInfo : SupportPowerInfo
 		[Desc("Weapon range offset to apply during the beacon clock calculation")]
 		public readonly WDist BeaconDistanceOffset = WDist.FromCells(6);
 
+		[Desc("How many attack runs to perform.")]
+		public readonly int Strikes = 1;
+
+		[Desc("How long to allow idling in the circle phase between strikes.")]
+		public readonly int CircleDelay = 30;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Magic value that is even unused (overriden with the same thing in YAML).

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs [hunk 2]
```diff
@@ -169,7 +175,13 @@ public void SendAirstrike(Actor self, WPos target, bool randomize = true, int at
 					attack.OnExitedAttackRange += onExitRange;
 					attack.OnRemovedFromWorld += onRemovedFromWorld;
 
-					a.QueueActivity(new Fly(a, Target.FromPos(target + spawnOffset)));
+					for (var strikes = 0; strikes < info.Strikes; strikes++)
+					{
+						a.QueueActivity(new Fly(a, Target.FromPos(target + spawnOffset)));
+						if (info.Strikes > 1)
+							a.QueueActivity(new FlyIdle(a, info.CircleDelay));
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Using `FlyTimed` here while keeping everything else the same actually produces much better results for the passes and the angles (provided the delay is still 30 ticks) that both look better and produces attack angles much closer to the original. I didn't verify if the angles are exactly like the original or the delay needs a bit of tweaking, but is probably close enough.

**Comment by penev92** [comment]
Unless someone is both willing and able to (because I'm willing, but I doubt the "able to" part) to implement proper formation flight when turning, I say we use that as an excuse to quickly resolve one of the bigger questions regarding this feature: *"Do we want to keep the original line formation or the ORA triangle formation?"*. Keeping the original formation is much, much easier and also true to the original, while using our "modern" formation would involve significantly more work.

So with that in mind, I suggest a couple of changes (which I might PR towards `Mailaender:3-orni-strikes` if we come to an agreement here) that would:
 - Give us the original line formation (also known as "drop-feeding your enemy" or "going in 1 by 1 to die easier")
 - Get us really close to the attack angles and circling area of the original (with the circling area being maybe slightly smaller, which is a good thing for the Ornithopters' survivability).

[This video of the original](https://gfycat.com/firmbetterkite) (that I linked above) produces [this bombing pattern](https://cdn.discordapp.com/attachments/632295260428566529/703706985182396466/unknown.png),
while [this video of ORA with my proposed changes to this PR](https://gfycat.com/warpedniftycuscus) produces [this bombing pattern](https://cdn.discordapp.com/attachments/632295260428566529/703707129655197706/unknown.png).
I think this would qualify as "close enough", as it looks pretty good to me.

The possibly only thing left to figure out is the plane's exit trajectory. In the original they exit in a straight line after their last bombing run, resulting in a 60° angle to their entry path (which starts at about the player's High Tech Factory), while here they try to exit in the same direction they faced when entering, which is a leftover of how Badgers and A10s work. I'm not sure this can be considered an issue, but it is a polishing thing.

Oh yes, the Camera actor for the support power is wired to only exist during the first bombing run. We should either keep it alive for the whole thing or recreate it for each pass.

@sorcerer86pt unfortunately your suggestion would run into issues at the first sign of enemy AA.

**Comment by pchote** [comment]
I agree with your solution.

**Comment by penev92** [comment]
This PR should also probably address [this comment in the weapon definition](https://github.com/OpenRA/OpenRA/blob/bleed/mods/d2k/weapons/other.yaml#L72).

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs (hunk 2 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Updated with your suggestions.

**Comment by Mailaender** [comment]
Done.

**Comment by abcdefg30** [comment]
Apparently this depends on #18007 to move forward.

File: mods/d2k/weapons/other.yaml [hunk 3]
```diff
@@ -69,7 +69,7 @@ OrniBomb:
 	Warhead@1Dam: SpreadDamage
 		Spread: 320
 		Falloff: 100, 60, 30, 15, 0
-		Damage: 7500 #400 in original, reduce when bombers can do multiple passes
+		Damage: 400
```
Comment (Porenutak) [inline mods/d2k/weapons/other.yaml:72]: Damage should be 4000 as ORA use 10x higher damage values.

**Comment by penev92** [comment]
It has been over two years since this PR was deemed insufficient to solve the task at hand and several tiers of dependencies/blockers came up and were shelves. There is no point in keeping it open indefinitely while we wait.

**Pull request closed without merging**